File: web/documentation/scripts/run_examples.py
#!/usr/bin/env python3
"""Run the Python documentation examples and cache their real output.

Reads ``examples-python.json``, executes each entry's ``code`` in a fresh
namespace, captures stdout into ``output``, and stamps ``captured_at``.

If an entry sets ``"plot"``, the code must leave the result object to be plotted
in a variable named ``result``; the runner renders it to
``../assets/examples/<plot>`` via that result's public ``save_plot`` (PNG, light-print
theme — the default). This script also doubles as the shared renderer: on every run it sweeps
``../assets/examples`` for any ``*.vl.json`` specs left by ``run_examples.R``
(the R port has no native PNG renderer) and converts them to PNG with
``vl_convert``, so both ports' charts come from one renderer. Run the R runner
first, then this one, to materialise every plot.

This fills the cache that ``inject_examples.py`` then pastes into the tutorial
pages — author ``code``, run this to fill ``output``/``captured_at`` and render
plots, then run ``inject_examples.py`` to update the pages. Nothing here is wired
into the leyline build; it is a one-shot cache refresher.

Usage:
    python run_examples.py            # refresh every entry
    python run_examples.py <id> ...   # refresh only the named entries
"""
from __future__ import annotations

import contextlib
import io
import json
import shutil
import sys
import tempfile
from datetime import datetime, timezone
from pathlib import Path

HERE = Path(__file__).resolve().parent
CACHE = HERE / "examples-python.json"
ASSETS = HERE.parent / "assets" / "examples"


def _render_spec_file(spec_path: Path) -> None:
    """Convert a Vega-Lite ``.vl.json`` spec to a same-stem PNG, then drop it."""
    import vl_convert as vlc

    png = spec_path.with_suffix("").with_suffix(".png")  # foo.vl.json -> foo.png
    png.write_bytes(vlc.vegalite_to_png(spec_path.read_text(), scale=2))
    spec_path.unlink()
    print(f"  rendered {png.name} (from R spec)")


def run_entry(entry: dict) -> None:
    namespace: dict = {}
    buf = io.StringIO()
    with contextlib.redirect_stdout(buf):
        exec(entry["code"], namespace)  # noqa: S102 — trusted, author-written
    entry["output"] = buf.getvalue().rstrip("\n")
    entry["captured_at"] = datetime.now(timezone.utc).isoformat(timespec="seconds")

    plot_name = entry.get("plot")
    if plot_name:
        result = namespace.get("result")
        if result is None:
            raise SystemExit(
                f"entry {entry['id']!r} sets 'plot' but left no 'result' variable"
            )
        ASSETS.mkdir(parents=True, exist_ok=True)
        with tempfile.TemporaryDirectory() as tmp:
            tmp_path = Path(tmp) / plot_name
            result.save_plot(str(tmp_path))
            shutil.copy2(tmp_path, ASSETS / plot_name)
        print(f"  plot: {plot_name}")


def main() -> None:
    entries = json.loads(CACHE.read_text())
    wanted = set(sys.argv[1:])
    for entry in entries:
        if wanted and entry["id"] not in wanted:
            continue
        print(f"running {entry['id']} ...")
        run_entry(entry)
    CACHE.write_text(json.dumps(entries, indent=2) + "\n")
    print(f"wrote {CACHE.name}")

    # Shared render step: turn any R-emitted specs into PNGs with one renderer.
    for spec_path in sorted(ASSETS.glob("*.vl.json")):
        _render_spec_file(spec_path)


if __name__ == "__main__":
    main()

File: web/documentation/scripts/run_examples.R
#!/usr/bin/env Rscript
# Run the R documentation examples and cache their real output.
#
# Mirror of run_examples.py for the R port. Reads examples-r.json, evaluates
# each entry's `code` capturing console output into `output`, and stamps
# `captured_at`.
#
# If an entry sets `plot`, the code must leave the result object to be plotted
# in a variable named `result`. The R port ships no built-in PNG renderer, so
# rather than render here this writes the themed Vega-Lite spec to
# ../assets/examples/<plot>.vl.json; `run_examples.py` then converts every such
# spec to PNG with vl_convert (one renderer for both ports). So: run this, then
# run run_examples.py to materialise the R plots.
#
# This fills the cache that inject_examples.py pastes into the tutorial pages:
# author `code`, run this to fill `output`/`captured_at` (and emit plot specs),
# then run inject_examples.py to update the pages. Not wired into the leyline
# build.
#
# Usage:
#   Rscript run_examples.R            # refresh every entry
#   Rscript run_examples.R id1 id2    # refresh only the named entries

suppressMessages({
  library(jsonlite)
  library(mcpower)
})

here   <- dirname(normalizePath(sub("--file=", "", grep("--file=", commandArgs(FALSE), value = TRUE))))
cache  <- file.path(here, "examples-r.json")
assets <- file.path(dirname(here), "assets", "examples")

# Write the primary Vega-Lite spec block for a result as a .vl.json file.
# Uses .plot_blocks() (the current internal API) to obtain the named block list,
# then picks the representative block (power for find_power, curve for
# find_sample_size). run_examples.py converts every .vl.json to PNG via
# vl_convert — so the R port ships no PNG renderer of its own.
# Fails gracefully with a warning when vegawidget/V8 are absent (they are not
# needed here — the spec is light-print-themed JSON written straight to disk).
write_plot_spec <- function(result, plot_name) {
  dir.create(assets, recursive = TRUE, showWarnings = FALSE)
  kind <- attr(result, "mcpower_kind")
  if (is.null(kind)) kind <- "find_power"
  blocks <- tryCatch(
    mcpower:::.plot_blocks(result, kind),
    error = function(e) {
      warning(sprintf("write_plot_spec: .plot_blocks() failed: %s", conditionMessage(e)))
      NULL
    }
  )
  if (is.null(blocks)) {
    return("(skipped — .plot_blocks() failed; see warning above)")
  }
  block_key <- if (kind == "find_sample_size") "curve" else "power"
  spec <- if (!is.null(blocks[[block_key]])) blocks[[block_key]] else blocks[[1L]]
  spec_json <- mcpower:::.apply_theme(
    toJSON(spec, auto_unbox = TRUE, null = "null"),
    "light-print"
  )
  spec_path <- file.path(assets, sub("\\.png$", ".vl.json", plot_name))
  writeLines(spec_json, spec_path)
  sprintf("(wrote light-print-themed spec %s; run_examples.py renders it to PNG)", basename(spec_path))
}

run_entry <- function(entry) {
  env <- new.env()
  out <- capture.output(eval(parse(text = entry$code), envir = env))
  entry$output <- paste(out, collapse = "\n")
  entry$captured_at <- format(Sys.time(), "%Y-%m-%dT%H:%M:%S%z")
  # plot is a single filename for chart entries; no-plot entries carry `null`,
  # which round-trips through jsonlite to `{}` (empty list) on re-run — guard
  # for a non-empty scalar string so both shapes are treated as "no plot".
  plot_name <- entry$plot
  if (is.character(plot_name) && length(plot_name) == 1L && nzchar(plot_name)) {
    result <- get0("result", envir = env)
    if (is.null(result)) stop(sprintf("entry %s sets 'plot' but left no 'result'", entry$id))
    message("  plot: ", write_plot_spec(result, plot_name))
  }
  entry
}

main <- function() {
  args    <- commandArgs(trailingOnly = TRUE)
  entries <- fromJSON(cache, simplifyVector = FALSE)
  for (i in seq_along(entries)) {
    if (length(args) && !(entries[[i]]$id %in% args)) next
    message("running ", entries[[i]]$id, " ...")
    entries[[i]] <- run_entry(entries[[i]])
  }
  writeLines(toJSON(entries, pretty = TRUE, auto_unbox = TRUE), cache)
  message("wrote ", basename(cache))
}

main()
